### PR TITLE
fix: dynamic parameters via sig_params

### PR DESCRIPTION
**Pull request by itsyoboieltr**
## **Description**

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/37086?quickstart=1)

Adds support for sig_params in deeplinks

## **Changelog**

CHANGELOG entry: null

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/6017

## **Manual testing steps**

1. sig_params should work according to the agreed spec

## **Screenshots/Recordings**

Not applicable

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> Introduce `sig_params` support in deep link canonicalization, parsing, signing, and tests to allow selectively signed/forwarded query params.
> 
> - **Deep Links**:
>   - **Canonicalization**: Enhance `canonicalize` to honor `sig_params` — keep only listed query params (preserving repeats), append `sig_params`, sort, and remove `sig`; fallback to previous behavior when `sig_params` is absent.
>   - **Parsing**: Use canonicalized URL, strip `sig_params` before invoking route handlers, and verify against the original URL.
>   - **Constants**: Add `SIG_PARAMS_PARAM` (`sig_params`).
> - **Tests**:
>   - **Unit**: Add cases for `sig_params` filtering, repetition preservation, non-mutation; adjust `parse` tests to pass empty params to handlers when only signature present.
>   - **E2E**: Expand scenarios for signed-with/without `sig_params` and unsigned; add assertions that only allowed params flow through; add `getHashParams` helper; update redirects to ignore query in expectations.
> - **Helpers**:
>   - Update `signDeepLink` to optionally append `sig_params` (derived from current params) and sign the full URL; add imports and sorting adjustments.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f8899798b0033f3ef97b58c2bfa8d02e1a074b7c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: shared/lib/deep-links/parse.ts [hunk 1]
```diff
@@ -18,7 +19,10 @@ export async function parse(url: URL): Promise<ParsedDeepLink | false> {
 
   let destination: Destination;
   try {
-    destination = route.handler(url.searchParams);
+    const params = new URLSearchParams(url.searchParams);
+    params.delete(SIG_PARAM);
+    params.delete(SIG_PARAMS);
+    destination = route.handler(params);
```
Comment (davidmurdoch) [inline shared/lib/deep-links/parse.ts]: all params that are not signed should be removed from the params passed to the handler.

File: shared/lib/deep-links/canonicalize.ts [hunk 2]
```diff
@@ -1,23 +1,44 @@
...
-  // clone the searchParams so we don't edit the original URL when deleting
-  // `sig`
-  const params = new URLSearchParams(url.searchParams);
-  params.delete(SIG_PARAM);
-  params.sort();
-  const queryString = params.toString();
-  const fullUrl =
-    url.origin + url.pathname + (queryString ? `?${queryString}` : '');
-  return fullUrl;
+  let queryString = '';
```
Comment (davidmurdoch) [inline shared/lib/deep-links/canonicalize.ts]: no need to assign `''` here, since its always overwritten later.

Comment (itsyoboieltr) [reply shared/lib/deep-links/canonicalize.ts]: fixed

File: shared/lib/deep-links/parse.ts (hunk 1 above)
Comment (itsyoboieltr) [reply shared/lib/deep-links/parse.ts]: is there anything stopping me from using canonicalize here?

File: shared/lib/deep-links/parse.test.ts [hunk 3]
```diff
@@ -70,7 +69,7 @@ describe('parse', () => {
       },
       signature: VALID,
     });
-    expect(mockHandler).toHaveBeenCalledWith(new URL(urlStr).searchParams);
+    expect(mockHandler).toHaveBeenCalledWith(new URLSearchParams());
```
Comment (itsyoboieltr) [inline shared/lib/deep-links/parse.test.ts:73]: This will become an empty URLSearchParams because we remove `sig`

File: shared/lib/deep-links/parse.ts (hunk 1 above)
Comment (itsyoboieltr) [reply shared/lib/deep-links/parse.ts]: NO!

File: shared/lib/deep-links/canonicalize.ts [hunk 4]
```diff
@@ -1,23 +1,44 @@
...
+  if (sigParams) {
+    const allowedParams = sigParams.split(',');
+    const signedParams = new URLSearchParams();
+
+    for (const allowedParam of allowedParams) {
+      const values = url.searchParams.getAll(allowedParam);
+      for (const value of values) {
+        signedParams.append(allowedParam, value);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (itsyoboieltr) [reply shared/lib/deep-links/canonicalize.ts:25]: sig_params is signed. If the attacker changes sig_params, the verification will fail. So cursor is wrong.

File: shared/lib/deep-links/canonicalize.ts [hunk 5]
```diff
@@ -1,23 +1,44 @@
...
-  const queryString = params.toString();
-  const fullUrl =
-    url.origin + url.pathname + (queryString ? `?${queryString}` : '');
-  return fullUrl;
+  let queryString: string | undefined;
+
+  const sigParams = url.searchParams.get(SIG_PARAMS);
+
+  if (sigParams) {
+    const allowedParams = sigParams.split(',');
```
[the remark this answers is not in the document]
Comment (itsyoboieltr) [reply shared/lib/deep-links/canonicalize.ts:17]: it literally doesn't contain spaces around commas

File: shared/lib/deep-links/constants.ts [hunk 6]
```diff
@@ -3,3 +3,4 @@
 export const DEEP_LINK_HOST = process.env.DEEP_LINK_HOST;
 export const DEEP_LINK_MAX_LENGTH = 2048;
 export const SIG_PARAM = 'sig';
+export const SIG_PARAMS = 'sig_params';
```
Comment (davidmurdoch) [inline shared/lib/deep-links/constants.ts]: `sig` is the name of the `param`, so `SIG_PARAMS` should be `SIG_PARAMS_param` 😅 

Or we rename to SIG_PARAM_NAME and `SIG_PARAMS_PARAM_NAME`?

File: shared/lib/deep-links/parse.ts [hunk 7]
```diff
@@ -18,7 +20,11 @@ export async function parse(url: URL): Promise<ParsedDeepLink | false> {
 
   let destination: Destination;
   try {
-    destination = route.handler(url.searchParams);
+    const canonicalUrl = new URL(canonicalize(url));
```
Comment (davidmurdoch) [inline shared/lib/deep-links/parse.ts:23]: `verify` also does `canonicalize(url)`.

So you _could_ just do:

```typescript
const canonicalUrl = canonicalize(url); // make canonicalize return a `URL` instead of a string
const signatureStr = url.searchParams.get("sig");
const signature = sig ? await verify(signatureStr, canonicalUrl) : MISSING;
try {
    canonicalUrl.delete(SIG_PARAMS);
    destination = route.handler(routeParams);
}
// ...
```

and update `verify` to:

a) require the `sig` in its function signature
b) no longer check if `sig` is present, since we can just do that here
c) no longer use `canonicalize`, as the URL is already `canonicalized`
d) accept a `URL` instead of a string (and then do this: `encoder.encode(canonicalUrl.toString())`)

This changes the order we process values. Previously we wouldn't call `verify` if the route wasn't valid. But in reality, I don't think this matters. I did it this way initially just to "fail fast", but I its not neccessary.

File: test/e2e/tests/deep-link/deep-link.spec.ts [hunk 8]
```diff
@@ -206,20 +206,15 @@ and we'll take you to the right place.`
         // test signed flow
```
Comment (davidmurdoch) [inline test/e2e/tests/deep-link/deep-link.spec.ts:214]: You need to update `signDeepLink` to support two types of _signed_ URLs: URLs _with_ `sig_params` and URLs without `sig_params`. Right now `signDeepLink` does not add `sig_params`, so we aren't testing that path.

An "easy" way to do this is to update the `cartesianProduct`'s `signed` values to something like `"signed with sig_params",  "signed without sig_params", "unsigned"`. Then update the test code to validate the two scenarios.

Additionally, we need to validate that params we don't want routes to have access to do not "leak" to the routes component. That'll just be a few new e2e test scenarios.

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: shared/lib/deep-links/parse.ts (hunk 7 above)
Comment (davidmurdoch) [reply shared/lib/deep-links/parse.ts:23]: (this is an _optional_ change, but if you don't do it, I will later :-) )

Comment (itsyoboieltr) [reply shared/lib/deep-links/parse.ts:23]: I agree that this can be done later

File: shared/lib/deep-links/constants.ts (hunk 6 above)
Comment (itsyoboieltr) [reply shared/lib/deep-links/constants.ts]: renamed to SIG_PARAMS_PARAM

File: test/e2e/tests/deep-link/helpers.ts [hunk 9]
```diff
@@ -15,24 +15,40 @@ export async function generateECDSAKeyPair() {
...
+  const signedUrl = new URL(canonicalUrl);
+
+  if (withSigParams) {
+    const sigParams = [...new Set(signedUrl.searchParams.keys())];
+
+    if (sigParams.length) {
+      signedUrl.searchParams.append('sig_params', sigParams.join(','));
+      signedUrl.searchParams.sort();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (itsyoboieltr) [reply test/e2e/tests/deep-link/helpers.ts:45]: it doesn't already include sig_params

File: test/e2e/tests/deep-link/deep-link.spec.ts (hunk 8 above)
Comment (itsyoboieltr) [reply test/e2e/tests/deep-link/deep-link.spec.ts:214]: added

File: test/e2e/tests/deep-link/deep-link.spec.ts [hunk 10]
```diff
@@ -113,17 +121,45 @@ describe('Deep Link', function () {
...
+              preparedUrl.searchParams.has('sig_params'),
+              false,
+              'signed without sig_params URLs must NOT include sig_params',
+            );
+            assert.equal(
+              preparedUrl.searchParams.has('sig'),
+              true,
+              'signed URLs must include sig',
+            );
+          }
```
Comment (davidmurdoch) [inline test/e2e/tests/deep-link/deep-link.spec.ts]: This only validates that the test's helper function `signDeepLink` works (I don't think this is necessary to validate here, since its only used in this one test file).

We _do_ need to validate that the page the browser ends up on _at the very end_ of the process has access to the correct params (that it has params it should have, and that it doesn't have params it should not have). That _doesn't_ have to be done in these `scenarios` tests though, if its easier to just create a new e2e `it` test specifically for that - as this scenario test is already pretty complicated.

Specifically, I think it is worth while to add regression tests for:

1. `?foo=0&foo=1&bar=2&baz=3&sig_params=foo,bar&sig=...` allows access to params `foo` (both) and `bar`, but _not_ `baz`
2. `?foo=1&bar=2&baz=3&sig=...` allows access to params `foo`, `bar`, and `baz`
3. `?foo=1&foo=2&bar=3` allows access to _both_ `foo` values _and_ the `bar` value (unsigned flow allows all)

I know we thought of a lot of possible edge cases here, so if you can think of more that might be worth testing you can add more tests for those too.

Comment (itsyoboieltr) [reply test/e2e/tests/deep-link/deep-link.spec.ts]: fixed

**Review by davidmurdoch — APPROVED** [review]

**Review by david0xd — APPROVED** [review]
